python/lsst/daf/butler/tests/_testRepo.py

```diff
@@ -388,32 +349,71 @@ def addDataIdValue(butler: Butler, dimension: str, value: Any, **related: Any):
-            "Could not create data ID value. Automatic relationship generation "
-            "may have failed; try adding keywords to assign a specific instrument, "
-            "physical_filter, etc. based on the nested exception message."
-        ) from e
+    # Assemble a dictionary data ID holding the given primary dimension value
+    # and all of the related ones.
+    data_id: dict[str, Union[int, str]] = {dimension: value}
+    data_id.update(related)
+
+    # Compute the set of all dimensions that these recursively depend on.
+    all_dimensions = butler.registry.dimensions.extract(data_id.keys())
+
+    # Create dicts that will become DimensionRecords for all of these data IDs.
+    # This iteration is guaranteed to be in topological order, so we can count
+    # on new data ID values being invented before they are needed.
+    record_dicts_by_dimension: dict[Dimension, dict[str, Any]] = {}
+    for dimension_obj in all_dimensions:
+        dimension_value = data_id.get(dimension_obj.name)
+        if dimension_value is None:
+            # No values given by caller for this dimension.  See if any exist
+            # in the registry that are consistent with the values of dimensions
+            # we do have:
+            match_data_id = {key: data_id[key] for key in data_id.keys() & dimension_obj.dimensions.names}
+            matches = list(
+                butler.registry.queryDimensionRecords(dimension_obj, dataId=match_data_id).limit(1)
+            )
+            if not matches:
+                # Nothing in the registry matches: invent a data ID value
+                # with the right type (actual value does not matter).
+                dimension_value = _makeRandomDataIdValue(dimension_obj)
+                data_id[dimension_obj.name] = dimension_value
+            else:
+                # A record does exist in the registry.  Use its data ID value,
+                # and skip the rest of the record-fabrication logic.
+                data_id[dimension_obj.name] = matches[0].dataId[dimension_obj.name]
+                continue
+        if dimension_obj.viewOf is not None:
+            # Don't need to bother generating full records for dimensions whose
+            # records are just a view into some other's records anyway.
+            continue
+        # Add the primary key field for this dimension.
+        record_dict: dict[str, Any] = {dimension_obj.primaryKey.name: dimension_value}
+        # Define secondary keys (e.g., detector name given detector id)
+        record_dict.update(_fillAllKeys(dimension_obj, dimension_value))
+        # Set the foreign key values for any related dimensions that should
+        # appear in the record.
+        for related_dimension in dimension_obj.dimensions:
+            if related_dimension.name != dimension_obj.name:
+                record_dict[related_dimension.name] = data_id[related_dimension.name]
+
+        record_dicts_by_dimension[dimension_obj] = record_dict
```

Comment:
Would it be simpler to just call `addDataIdValue` recursively for missing required dimensions? I'm worried about `syncDimensionData` either failing, or worse, changing, records for dimensions that existed before the call.

Reply:
`syncDimensionData` definitely will not change existing records, and I think if it does fail it does still say something important.  That said, I'm not sure I've left it possible for it to fail - I have not been able to trigger the previous expected failures in the tests without breaking something else (more on that in other threads).

Reply:
I did look a bit into calling `addDataIdValue` recursively just now, and I think it'd a bit messier in the end, because we'd have to add new arguments to distinguish between the original contents of `**related` and later derived ones.